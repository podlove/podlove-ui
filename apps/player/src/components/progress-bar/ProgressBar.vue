<template>
  <progress-bar
    :progress-color="progressColor"
    :thumb-color="state.thumbColor"
    :highlight-color="state.highlightColor"
    :duration="state.duration"
    :time="state.playtime"
    :ghost="state.ghost"
    :buffer="state.buffer"
    :chapters="state.chapters"
    :quantiles="state.quantiles"
    :title="$t(state.title.key, state.title.attr)"
    data-test="progress-bar"
    @time="dispatch"
    @simulate="dispatch"
    @ghost="dispatch"
  />
</template>

<script>
import { mapState, injectStore } from 'redux-vuex'
import { fade } from 'farbraum'
import ProgressBar from '@podlove/components/progress-bar/ProgressBar.vue'

import select from '../../store/selectors'

export default {
  components: {
    ProgressBar
  },
  setup() {
    return {
      state: mapState({
        progressColor: select.theme.brandDark,
        thumbColor: select.theme.brandDark,
        highlightColor: select.theme.brandLightest,
        duration: select.duration,
        playtime: select.playtime,
        ghost: select.ghost.time,
        buffer: select.network.buffer,
        chapters: select.chapters.list,
        quantiles: select.quantiles,
        title: select.accessibility.progressBar
      }),
      dispatch: injectStore().dispatch
    }
  },
  computed: {
    progressColor() {
      return fade(this.state.progressColor, 0.7)
    }
  }
}
</script>
