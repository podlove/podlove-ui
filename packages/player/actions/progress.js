import { ENABLE_GHOST_MODE, DISABLE_GHOST_MODE } from './types'
import { createAction } from './action-creator'

export const enableGhost = createAction(ENABLE_GHOST_MODE)
export const disableGhost = createAction(DISABLE_GHOST_MODE)
