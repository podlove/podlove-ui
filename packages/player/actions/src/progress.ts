import { ENABLE_GHOST_MODE, DISABLE_GHOST_MODE } from './types.js';
import { createAction } from 'redux-actions';

export type enableGhostPayload = void;
export type disableGhosttPayload = void;

export const enableGhost = createAction<enableGhostPayload>(ENABLE_GHOST_MODE);
export const disableGhost = createAction<disableGhosttPayload>(DISABLE_GHOST_MODE);
