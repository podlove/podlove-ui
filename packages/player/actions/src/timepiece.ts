import { createAction } from 'redux-actions';

import {
  BACKEND_PLAYTIME,
  REQUEST_PLAYTIME,
  SIMULATE_PLAYTIME,
  BACKEND_DURATION,
  BACKEND_LIVESYNC,
  UPDATE_PLAYTIME
} from './types.js';

export type backendPlaytimePayload = number;
export type requestPlaytimePayload = number;
export type simulatePlaytimePayload = number;
export type backendDurationPayload = number;
export type backendLiveSyncPayload = number;
export type updatePlaytimePayload = number;

export const backendPlaytime = createAction<backendPlaytimePayload>(BACKEND_PLAYTIME);
export const requestPlaytime = createAction<requestPlaytimePayload>(REQUEST_PLAYTIME);
export const simulatePlaytime = createAction<simulatePlaytimePayload>(SIMULATE_PLAYTIME);
export const backendDuration = createAction<backendDurationPayload>(BACKEND_DURATION);
export const backendLiveSync = createAction<backendLiveSyncPayload>(BACKEND_LIVESYNC);
export const updatePlaytime = createAction<updatePlaytimePayload>(UPDATE_PLAYTIME);
