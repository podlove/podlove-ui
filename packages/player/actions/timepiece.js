import { createAction } from './action-creator'

import {
  BACKEND_PLAYTIME,
  REQUEST_PLAYTIME,
  SIMULATE_PLAYTIME,
  BACKEND_DURATION,
  BACKEND_LIVESYNC
} from './types'

export const backendPlaytime = createAction(BACKEND_PLAYTIME)
export const requestPlaytime = createAction(REQUEST_PLAYTIME)
export const simulatePlaytime = createAction(SIMULATE_PLAYTIME)
export const backendDuration = createAction(BACKEND_DURATION)
export const backendLiveSync = createAction(BACKEND_LIVESYNC)
