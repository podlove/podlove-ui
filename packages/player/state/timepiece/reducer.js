import { handleActions } from 'redux-actions'
import {
  READY,
  BACKEND_DURATION,
  BACKEND_PLAYTIME,
  REQUEST_PLAYTIME,
  BACKEND_LIVESYNC
} from '@podlove/player-actions/types'
import { toInt } from '@podlove/utils/helper'
import { duration, playtime } from '@podlove/player-config'

export const INITIAL_STATE = {
  duration: 0,
  playtime: 0,
  livesync: 0
}

export const reducer = handleActions(
  {
    [READY]: (state, { payload }) => ({
      ...state,
      duration: duration(payload),
      playtime: playtime(payload)
    }),
    [BACKEND_DURATION]: (state, { payload }) => ({
      ...state,
      duration: toInt(payload || state.duration)
    }),
    [BACKEND_PLAYTIME]: (state, { payload }) => ({
      ...state,
      playtime: toInt(payload)
    }),
    [REQUEST_PLAYTIME]: (state, { payload }) => ({
      ...state,
      playtime: toInt(payload)
    }),
    [BACKEND_LIVESYNC]: (state, { payload }) => ({
      ...state,
      livesync: toInt(payload)
    })
  },
  INITIAL_STATE
)
