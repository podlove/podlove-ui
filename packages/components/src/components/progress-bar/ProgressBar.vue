<template>
  <div class="podlove-component--progress-bar w-full relative cursor-pointer">
    <input
      v-if="isMobile"
      ref="inputElement"
      type="range"
      min="0"
      :max="interpolate(duration)"
      :value="interpolate(time)"
      tabindex="-1"
      aria-hidden="true"
      @change="onChange(extractValue($event))"
      @input="onInput(extractValue($event))"
    />
    <input
      v-else
      ref="inputElement"
      type="range"
      min="0"
      :max="interpolate(duration)"
      :value="interpolate(time)"
      tabindex="-1"
      aria-hidden="true"
      @change="onChange(extractValue($event))"
      @input="onInput(extractValue($event))"
      @mousemove="onMouseMove"
      @mouseout="onMouseOut"
    />
    <span class="progress-range block absolute w-full left-0 pointer-events-none" />
    <span
      v-for="(buffering, index) in buffer"
      :key="`buffer-${index}`"
      tabindex="-1"
      aria-hidden="true"
      class="progress-buffer"
      :style="bufferStyle(buffering)"
    />
    <span
      v-for="(quantile, index) in quantiles"
      :key="`track-${index}`"
      tabindex="-1"
      aria-hidden="true"
      class="progress-track block absolute left-0 pointer-events-none"
      :style="trackStyle(quantile)"
    />
    <div>
      <span
        v-for="(chapter, index) in chapters"
        :key="index"
        tabindex="-1"
        aria-hidden="true"
        data-test="progress-bar--chapter-progress--indicator"
        class="chapters-progress absolute pointer-events-none"
        :class="{
          active: chapter.start <= ghost && chapter.end >= ghost
        }"
        :style="chapterStyle(chapter)"
      />
    </div>
    <span
      class="ghost-thumb absolute border-transparent border bg-opacity-75 pointer-events-none"
      :class="{
        hidden: !ghost,
        block: ghost
      }"
      aria-hidden="true"
      :style="ghostStyle"
    />
    <span
      class="progress-thumb absolute block border border-solid pointer-events-none"
      tabindex="-1"
      aria-hidden="true"
      :class="{ active: thumbActive }"
      :style="thumbStyle"
    />
    <input
      type="range"
      class="sr-only"
      :title="title"
      min="0"
      max="100"
      steps="1"
      :value="(time / duration) * 100"
      @change="onChange((extractValue($event) / 100) * duration)"
      @input="onInput((extractValue($event) / 100) * duration)"
    />
  </div>
</template>

<script lang="ts" setup>
import RangeTouch from 'rangetouch';

import { interpolate, relativePosition } from '@podlove/utils/math';
import { isMobile } from '@podlove/utils/detect';
import { requestPlaytime, simulatePlaytime } from '@podlove/player-actions/timepiece';
import { enableGhost, disableGhost } from '@podlove/player-actions/progress';

import { computed, onMounted, ref } from 'vue';
import { pathOr, path } from 'ramda';

const props = defineProps({
  time: {
    type: Number,
    default: 0
  },
  duration: {
    type: Number,
    default: 0
  },
  ghost: {
    type: Number,
    default: undefined
  },
  buffer: {
    type: Array,
    default: () => []
  },
  quantiles: {
    type: Array,
    default: () => []
  },
  title: {
    type: String,
    default: ''
  },
  chapters: {
    type: Array,
    default: () => []
  }
});

const emits = defineEmits(['time', 'ghost', 'simulate']);

const thumbStyle = computed(() => ({
  left: relativePosition(props.time, props.duration)
}));

const ghostStyle = computed(() => ({
  left: relativePosition(props.ghost, props.duration)
}));

const thumbActive = ref(false);

const inputElement = ref(null);

onMounted(() => {
  new RangeTouch(inputElement, { watch: false });
});

// Methods
const onChange = (value: number) => emits('time', requestPlaytime(value));
const onInput = (value: number) => {
  emits('time', requestPlaytime(value));
  emits('ghost', disableGhost());
};

const extractValue = (input: Event): number => path(['target', 'value'], input);

const onMouseMove = (event: MouseEvent) => {
  const width = pathOr(0, ['target', 'clientWidth'], event);

  if ((event.offsetY < 13 && event.offsetY > 31) || event.offsetX < 0 || event.offsetX > width) {
    thumbActive.value = false;
    emits('ghost', disableGhost());
    return false;
  }

  thumbActive.value = true;

  emits('simulate', simulatePlaytime((props.duration * event.offsetX) / width));
  emits('ghost', enableGhost());

  return false;
};

const onMouseOut = () => {
  thumbActive.value = false;

  emits('ghost', disableGhost());
  emits('simulate', simulatePlaytime(props.time));

  return false;
};

const trackStyle = ([start, end]: [number, number]) => ({
  left: relativePosition(start, props.duration),
  width: relativePosition(end - start, props.duration)
});

const bufferStyle = ([start, end]: [number, number]) => ({
  left: relativePosition(start, props.duration),
  width: relativePosition(end - start, props.duration)
});

const chapterStyle = (chapter: { start: number; end: number }) => ({
  left: (chapter.start * 100) / props.duration + '%',
  width: ((chapter.end - chapter.start) * 100) / props.duration + '%'
});
</script>

<style lang="scss" scoped>
@import '../../styles/range';

.podlove-component--progress-bar {
  @include range($progress-height, $thumb-width-desktop, $thumb-width-desktop-hover);
  height: $progress-height;
  transition: opacity 150ms, height 300ms;

  .progress-range {
    top: calc(50% - calc($track-height / 2));
    height: $track-height;
    background-color: var(
      --podlove-component--progress-bar--progress-color,
      var(--podlove-components-text)
    );
  }

  .chapters-progress {
    background: transparent;
    height: 2px;
    top: calc(50% - 1px);
    border-right: 2px solid
      var(
        --podlove-component--progress-bar--chapters-separator-color,
        var(--podlove-components-highlight)
      );

    &.active {
      height: 4px;
      top: calc(50% - 2px);

      background: var(
        --podlove-component--progress-bar--chapters-background-color,
        var(--podlove-components-text)
      );
    }
  }

  .progress-track {
    top: calc(50% - calc($track-height / 2));
    height: $track-height;
    background-color: var(
      --podlove-component--progress-bar--track-background-color,
      var(--podlove-components-text)
    );
  }

  .progress-thumb {
    margin-left: calc(-1px - calc($thumb-size / 2));
    height: $thumb-size;
    border-radius: $thumb-size;
    top: calc(50% - calc($thumb-size / 2));
    width: $thumb-size;

    transition: left 150ms;
    background-color: var(
      --podlove-component--progress-bar--thumb-background-color,
      var(--podlove-components-text)
    );
    border-color: var(
      --podlove-component--progress-bar--thumb-border-color,
      var(--podlove-components-highlight)
    );

    &.active {
      width: $thumb-active-size;
      height: $thumb-active-size;
      border-radius: $thumb-active-size;
      top: calc(50% - calc($thumb-active-size / 2));
    }
  }

  .ghost-thumb {
    margin-left: calc(-1px - calc($thumb-size / 2));
    height: $thumb-size;
    border-radius: $thumb-size;
    top: calc(50% - calc($thumb-size / 2));
    width: $thumb-size;
    background-color: var(
      --podlove-component--progress-bar--ghost-background-color,
      var(--podlove-components-text)
    );
    border-color: var(
      --podlove-component--progress-bar--ghost-border-color,
      var(--podlove-components-highlight)
    );
  }

  .progress-buffer {
    display: block;
    opacity: 1;
    position: absolute;
    height: $track-height;
    top: calc(50% - calc($track-height / 2));
    left: 0;
    pointer-events: none;
    background-color: var(
      --podlove-component--progress-bar--buffer-background-color,
      var(--podlove-components-highlight)
    );
  }
}
</style>
