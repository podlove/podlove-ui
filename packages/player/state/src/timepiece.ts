import { prop } from 'ramda';
import { Action, handleActions } from 'redux-actions';
import { toInt } from '@podlove/utils/helper';
import { duration, playtime } from '@podlove/player-config';
import { ready } from '@podlove/player-actions/lifecycle';

import { readyPayload } from '@podlove/player-actions/lifecycle';
import { backendDuration } from '@podlove/player-actions/timepiece';
import { backendDurationPayload } from '@podlove/player-actions/timepiece';
import { backendPlaytime } from '@podlove/player-actions/timepiece';
import { backendPlaytimePayload } from '@podlove/player-actions/timepiece';
import { requestPlaytime } from '@podlove/player-actions/timepiece';
import { requestPlaytimePayload } from '@podlove/player-actions/timepiece';
import { backendLiveSync } from '@podlove/player-actions/timepiece';
import { backendLiveSyncPayload } from '@podlove/player-actions/timepiece';

export interface State {
  duration: number;
  playtime: number;
  livesync: number;
}

export const INITIAL_STATE: State = {
  duration: 0,
  playtime: 0,
  livesync: 0
};

export const reducer = handleActions<State, readyPayload | backendDurationPayload | backendPlaytimePayload | requestPlaytimePayload | backendLiveSyncPayload>(
  {
    [ready.toString()]: (state, { payload }: Action<readyPayload>) => ({
      ...state,
      duration: duration(payload),
      playtime: playtime(payload)
    }),
    [backendDuration.toString()]: (state, { payload }: Action<backendDurationPayload>) => ({
      ...state,
      duration: toInt(payload || state.duration)
    }),
    [backendPlaytime.toString()]: (state, { payload }: Action<backendPlaytimePayload>) => ({
      ...state,
      playtime: toInt(payload)
    }),
    [requestPlaytime.toString()]: (state, { payload }: Action<requestPlaytimePayload>) => ({
      ...state,
      playtime: toInt(payload)
    }),
    [backendLiveSync.toString()]: (state, { payload }: Action<backendLiveSyncPayload>) => ({
      ...state,
      livesync: toInt(payload)
    })
  },
  INITIAL_STATE
);

export const selectors = {
  playtime: prop('playtime') as (input: State) => number,
  duration: prop('duration') as (input: State) => number,
  livesync: prop('livesync') as (input: State) => number,
}
