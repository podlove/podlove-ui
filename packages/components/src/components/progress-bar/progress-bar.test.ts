import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import { requestPlaytime } from '@podlove/player-actions/timepiece';
import { disableGhost } from '@podlove/player-actions/progress';

import ProgressBar from './ProgressBar.vue';

describe('ProgressBar', () => {
  describe('component', () => {
    test('should be a Vue instance', async () => {
      const wrapper = await mount(ProgressBar);

      expect(wrapper.vm).toBeTruthy();
    });
  });

  describe('props', () => {
    test('should render with default props', async () => {
      const wrapper = await mount(ProgressBar);

      expect(wrapper.element).toMatchSnapshot();
    });

    const props = [
      {
        prop: 'time',
        value: 5000
      },
      {
        prop: 'ghost',
        value: 30
      },
      {
        prop: 'buffer',
        value: [
          [0, 1000],
          [5000, 6000]
        ]
      },
      {
        prop: 'quantiles',
        value: [
          [0, 300],
          [5500, 6000]
        ]
      },
      {
        prop: 'title',
        value: 'title'
      },
      {
        prop: 'chapters',
        value: [1000, 5000]
      }
    ];

    props.forEach(({ prop, value }) => {
      test(`should render '${prop}'`, async () => {
        const wrapper = await mount(ProgressBar, {
          props: {
            [prop]: value,
            duration: 10000
          }
        });

        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('events', () => {
    test('should emit REQUEST_PLAYTIME on input event', async () => {
      const wrapper = await mount(ProgressBar, {
        props: {
          duration: 10000
        }
      });

      await wrapper.find('input').trigger('input');
      expect(wrapper.emitted().ghost).toEqual([[disableGhost()]]);
      expect(wrapper.emitted().time).toEqual([[requestPlaytime(0)]]);
    });

    test('should emit REQUEST_PLAYTIME on input event', async () => {
      const wrapper = await mount(ProgressBar, {
        props: {
          duration: 10000
        }
      });

      await wrapper.find('input').trigger('input');

      expect(wrapper.emitted().ghost).toEqual([[disableGhost()]]);
      expect(wrapper.emitted().time).toEqual([[requestPlaytime(0)]]);
    });

    test('should emit DISABLE_GHOST on input event', async () => {
      const wrapper = await mount(ProgressBar, {
        props: {
          duration: 10000
        }
      });

      await wrapper.find('input').trigger('input');

      expect(wrapper.emitted().ghost).toEqual([[disableGhost()]]);
      expect(wrapper.emitted().time).toEqual([[requestPlaytime(0)]]);
    });
  });
});
