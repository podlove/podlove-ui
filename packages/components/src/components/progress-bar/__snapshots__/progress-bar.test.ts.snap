// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressBar > props > should render with custom props 1`] = `
<div
  class="podlove-component--progress-bar w-full relative cursor-pointer"
  data-v-0f66235d=""
>
  <span
    class="progress-range block absolute w-full left-0"
    data-v-0f66235d=""
  />
  
  <span
    aria-hidden="true"
    class="progress-buffer pointer-events-none"
    data-v-0f66235d=""
    style="left: 0%; width: 10%;"
    tabindex="-1"
  />
  <span
    aria-hidden="true"
    class="progress-buffer pointer-events-none"
    data-v-0f66235d=""
    style="left: 50%; width: 10%;"
    tabindex="-1"
  />
  
  
  <span
    aria-hidden="true"
    class="progress-track block absolute left-0 pointer-events-none"
    data-v-0f66235d=""
    style="left: 0%; width: 3%;"
    tabindex="-1"
  />
  <span
    aria-hidden="true"
    class="progress-track block absolute left-0 pointer-events-none"
    data-v-0f66235d=""
    style="left: 55%; width: 5%;"
    tabindex="-1"
  />
  
  <div
    data-v-0f66235d=""
  >
    
    <span
      aria-hidden="true"
      class="chapters-progress absolute pointer-events-none active"
      data-test="progress-bar--chapter-progress--indicator"
      data-v-0f66235d=""
      style="left: 0%; width: 10%;"
      tabindex="-1"
    />
    <span
      aria-hidden="true"
      class="chapters-progress absolute pointer-events-none"
      data-test="progress-bar--chapter-progress--indicator"
      data-v-0f66235d=""
      style="left: 10%; width: 40%;"
      tabindex="-1"
    />
    <span
      aria-hidden="true"
      class="chapters-progress absolute pointer-events-none"
      data-test="progress-bar--chapter-progress--indicator"
      data-v-0f66235d=""
      style="left: 50%; width: 50%;"
      tabindex="-1"
    />
    
  </div>
  <span
    aria-hidden="true"
    class="ghost-thumb absolute border-transparent border bg-opacity-75 pointer-events-none block"
    data-v-0f66235d=""
    style="left: 0.3%;"
  />
  <span
    aria-hidden="true"
    class="progress-thumb absolute block border border-solid pointer-events-none"
    data-v-0f66235d=""
    style="left: 50%;"
    tabindex="-1"
  />
  <input
    class="sr-only"
    data-v-0f66235d=""
    max="100"
    min="0"
    steps="1"
    title="title"
    type="range"
    value="50"
  />
  <input
    aria-hidden="true"
    data-v-0f66235d=""
    max="10000"
    min="0"
    tabindex="-1"
    type="range"
    value="5000"
  />
</div>
`;

exports[`ProgressBar > props > should render with default props 1`] = `
<div
  class="podlove-component--progress-bar w-full relative cursor-pointer"
  data-v-0f66235d=""
>
  <span
    class="progress-range block absolute w-full left-0"
    data-v-0f66235d=""
  />
  
  
  
  
  <div
    data-v-0f66235d=""
  >
    
    
  </div>
  <span
    aria-hidden="true"
    class="ghost-thumb absolute border-transparent border bg-opacity-75 pointer-events-none hidden"
    data-v-0f66235d=""
    style="left: 100%;"
  />
  <span
    aria-hidden="true"
    class="progress-thumb absolute block border border-solid pointer-events-none"
    data-v-0f66235d=""
    style="left: 100%;"
    tabindex="-1"
  />
  <input
    class="sr-only"
    data-v-0f66235d=""
    max="100"
    min="0"
    steps="1"
    title=""
    type="range"
    value="NaN"
  />
  <input
    aria-hidden="true"
    data-v-0f66235d=""
    max="0"
    min="0"
    tabindex="-1"
    type="range"
    value="0"
  />
</div>
`;
